ext/SummationByPartsOperatorsOptimForwardDiffExt.jl

```diff
@@ -0,0 +1,239 @@
+    basis_functions_orthonormalized, basis_functions_orthonormalized_derivatives = orthonormalize_gram_schmidt(basis_functions,
+                                                                                                               basis_functions_derivatives,
+                                                                                                               nodes)
+    V = vandermonde_matrix(basis_functions_orthonormalized, nodes)
+    V_x = vandermonde_matrix(basis_functions_orthonormalized_derivatives, nodes)
+    # Here, W satisfies W'*W = I
+    # W = [V; -V_x]
+
+    B = spzeros(eltype(nodes), N, N)
+    B[1, 1] = -1
+    B[N, N] = 1
+
+    R = B * V / 2
+    x_length = x_max - x_min
+    S = zeros(eltype(nodes), N, N)
+    A = zeros(eltype(nodes), N, K)
+    SV = zeros(eltype(nodes), N, K)
+    PV_x = zeros(eltype(nodes), N, K)
+    S_cache = DiffCache(S)
+    A_cache = DiffCache(A)
+    SV_cache = DiffCache(SV)
+    PV_x_cache = DiffCache(PV_x)
+    daij_dsigmak = zeros(eltype(nodes), N, K, L)
+    daij_drhok = zeros(eltype(nodes), N, K, N)
+    p = (V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok)
+    # @views function optimization_function(x, p)
+    #     V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache = p
+    #     S = get_tmp(S_cache, x)
+    #     A = get_tmp(A_cache, x)
+    #     SV = get_tmp(SV_cache, x)
+    #     PV_x = get_tmp(PV_x_cache, x)
+    #     (N, _) = size(R)
+    #     L = Integer(N*(N - 1)/2)
+    #     sigma = x[1:L]
+    #     rho = x[(L + 1):end]
+    #     set_S!(S, sigma, N)
+    #     P = create_P(rho, x_length)
+    #     mul!(SV, S, V)
+    #     mul!(PV_x, P, V_x)
+    #     @. A = SV - PV_x + R
+    #     # Use the Frobenius norm since it is strictly convex and cheap to compute
+    #     return norm(A)^2
+    # end
+    @views function optimization_function_and_grad!(F, G, x, p)
+        V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok = p
+        S = get_tmp(S_cache, x)
+        A = get_tmp(A_cache, x)
+        SV = get_tmp(SV_cache, x)
+        PV_x = get_tmp(PV_x_cache, x)
+        (N, _) = size(R)
+        L = Integer(N*(N - 1)/2)
+        sigma = x[1:L]
+        rho = x[(L + 1):end]
+        set_S!(S, sigma, N)
+        P = create_P(rho, x_length)
+        mul!(SV, S, V)
+        mul!(PV_x, P, V_x)
+        @. A = SV - PV_x + R
+        if !isnothing(G)
+            fill!(daij_dsigmak, zero(eltype(daij_dsigmak)))
+            for i in 1:N
+                for j in 1:K
+                    for k in 1:L
+                        l_tilde = k + i - N * (i - 1) + div(i * (i - 1), 2)
+                        # same as above, but needs more type conversions
+                        # l_tilde = Int(k + i - (i - 1) * (N - i/2))
+                        if i + 1 <= l_tilde <= N
+                            daij_dsigmak[i, j, k] += V[l_tilde, j]
```

Comment:
Can you change the order of the loops? Julia is column-major, so the `i` index of `daij_dsigmak` should be updated closely together

Reply:
Done in [c133295](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/c133295f4e1cab0c79efe2ba92c34cb232d6e7dd).